torch/nn/modules/batchnorm.py

```diff
@@ -8,43 +8,56 @@
 # TODO: use separate backend functions?
 class _BatchNorm(Module):
 
-    def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=True):
+    def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=True,
+                 track_running_stats=True):
         super(_BatchNorm, self).__init__()
         self.num_features = num_features
-        self.affine = affine
         self.eps = eps
         self.momentum = momentum
+        self.affine = affine
+        self.track_running_stats = track_running_stats
         if self.affine:
             self.weight = Parameter(torch.Tensor(num_features))
             self.bias = Parameter(torch.Tensor(num_features))
         else:
             self.register_parameter('weight', None)
             self.register_parameter('bias', None)
-        self.register_buffer('running_mean', torch.zeros(num_features))
-        self.register_buffer('running_var', torch.ones(num_features))
+        if self.track_running_stats:
+            self.register_buffer('running_mean', torch.zeros(num_features))
+            self.register_buffer('running_var', torch.ones(num_features))
+        else:
+            self.register_parameter('running_mean', None)
+            self.register_parameter('running_var', None)
         self.reset_parameters()
 
     def reset_parameters(self):
-        self.running_mean.zero_()
-        self.running_var.fill_(1)
+        if self.track_running_stats:
+            self.running_mean.zero_()
+            self.running_var.fill_(1)
         if self.affine:
             self.weight.data.uniform_()
             self.bias.data.zero_()
 
+    def _check_input_dim(self, input):
+        return NotImplemented
+
     def forward(self, input):
+        self._check_input_dim(input)
+
         return F.batch_norm(
             input, self.running_mean, self.running_var, self.weight, self.bias,
-            self.training, self.momentum, self.eps)
+            self.training or not self.track_running_stats, self.momentum, self.eps)
 
     def __repr__(self):
         return ('{name}({num_features}, eps={eps}, momentum={momentum},'
-                ' affine={affine})'
+                ' affine={affine}, track_running_stats={track_running_stats})'
                 .format(name=self.__class__.__name__, **self.__dict__))
 
 
 class BatchNorm1d(_BatchNorm):
-    r"""Applies Batch Normalization over a 2d or 3d input that is seen as a
-    mini-batch.
+    r"""Applies Batch Normalization over a 3d input (a mini-batch of 2d inputs)
```

Comment:
Is this not still 2D and 3D? This was overloaded to do the original batch norm with N x C (e.g. fully-connected layers) and the 1D signal version N x C x L (e.g. 1D conv layers).